aineko/extras/connectors/wss.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2023 Aineko Authors
+# SPDX-License-Identifier: Apache-2.0
+"""Extra module for connecting to a WebSocket."""
+
+import json
+import time
+from typing import Any, Dict, List, Optional
+
+import websocket
+from pydantic import BaseModel, field_validator
+
+from aineko import AbstractNode
+from aineko.utils.secrets import inject_secrets
+
+
+class ParamsWSS(BaseModel):
+    """Connector params for WebSocket model."""
+
+    max_retries: int = 30
+    retry_sleep: float = 5
+    url: str
+    header: Optional[Dict[str, str]] = None
+    init_messages: List[Any] = []
+    metadata: Optional[Dict[str, Any]] = None
+
+    @field_validator("url")
+    def supported_url(cls, u: str) -> str:  # pylint: disable=no-self-argument
+        """Validates that the url is a valid WebSocket URL."""
+        if not u.startswith("wss://") or u.startswith("ws://"):
+            raise ValueError(
+                "Invalid url provided to WebSocket params. "
+                'Expected url to start with "wss://" or "ws://". '
+                f"Provided url was: {str(u)}"
+            )
+        return u
+
+
+class WSS(AbstractNode):
+    """Connector for a websocket."""
+
+    retry_count = 0
+
+    def _pre_loop_hook(self, params: dict | None = None) -> None:
+        """Initalize the WebSocket connection."""
+        # Cast params to ParamsWSS type
+        try:
+            if params is not None:
+                self.wss_params = ParamsWSS(**params)
+            else:
+                raise ValueError(
+                    "No params provided to WSS connector. "
+                    "Node requires at least a url param."
+                )
+        except Exception as err:  # pylint: disable=broad-except
```

Comment:
I would personally add more specific error handling than just the broad error. Maybe a websocket specific error.

Reply:
This logic is necessary since ray cannot pickle all exceptions. So the workaround is to catch all exceptions and raise an exception that ray can pickle, while casting the error to a str so you don't lose information about the error. in this case the possible excpetions are pydantic exceptions, which are known to be non-picklable.

I'll add some documentation for this.

Reply:
https://github.com/aineko-dev/aineko/pull/121/commits/faf2d9754b685ce6a87e0b10380a9d6abef8e481